kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceRateLimiter.java

```diff
@@ -98,29 +105,48 @@ private boolean isOverGracePeriod(Long nowMs) {
     return false;
   }
 
-  private int addAndGetRate(long nowMs) {
-    rateCounter.add(nowMs);
+  private void addToRateLimiter(TimeAndSize thisMessage) {
+    rateCounter.add(thisMessage);
     rateCounterSize.incrementAndGet();
+    byteCounterSize.addAndGet(thisMessage.size);
 
     synchronized (rateCounter) {
-      if (rateCounter.peekLast() < nowMs - ONE_SECOND_MS) {
+      if ((rateCounter.peekLast().time < thisMessage.time - ONE_SECOND_MS)) {
         rateCounter.clear();
         rateCounterSize.set(0);
       } else {
-        while (rateCounter.peek() < nowMs - ONE_SECOND_MS) {
-          rateCounter.poll();
+        while (rateCounter.peek().time < thisMessage.time - ONE_SECOND_MS) {
+          TimeAndSize messageToRemove = rateCounter.poll();
+          byteCounterSize.addAndGet(-messageToRemove.size);
           rateCounterSize.decrementAndGet();
         }
       }
     }
-    return rateCounterSize.get();
+    return;
   }
 
-  private Optional<Duration> getWaitFor(int currentRate) {
-    if (currentRate <= maxRequestsPerSecond) {
+  private Optional<Duration> getWaitFor(int currentCountRate, long currentByteRate) {
+    if (currentCountRate <= maxRequestsPerSecond && currentByteRate <= maxBytesPerSecond) {
       return Optional.empty();
     }
-    double waitForMs = ((double) currentRate / (double) maxRequestsPerSecond - 1) * 1000;
+
+    double waitForMs;
+    if (currentCountRate > maxRequestsPerSecond) {
+      waitForMs = ((double) currentCountRate / (double) maxRequestsPerSecond - 1) * 1000;
+    } else {
```

Comment:
Shouldn't we check if `currentByteRate > maxBytesPerSecond`? Or else we could end up with a negative `waitForMs`?

Reply:
I didn't think we needed to

If currentCountRate !> maxRequestsPerSecond then we hit the else statement covering the byte rate, but we can be sure that currentByteRate > maxBytesPerSecond (and so no negative wait for) because of the check on line 129 :

(currentCountRate <= maxRequestsPerSecond && currentByteRate <= maxBytesPerSecond)

One of count rate or byte rate is over their limits, and if it's not count rate, then it must be byte rate, so no need to recheck.

Reply:
Ow, sorry, I missed line 129. Ack. :)